pumpkin/src/block/blocks/crafting_table.rs

```diff
@@ -6,52 +6,112 @@ use async_trait::async_trait;
+            .await;
         BlockActionResult::Consume
     }
+
+    async fn on_broken<'a>(
+        &self,
+        block: &Block,
+        player: &Player,
+        location: WorldPosition,
+        server: &Server,
+    ) {
+        super::standard_on_destroy_with_container(block, player, location, server).await;
+    }
+
+    async fn on_close<'a>(
+        &self,
+        _block: &Block,
+        player: &Player,
+        _location: WorldPosition,
+        _server: &Server,
+        container: &OpenContainer,
+    ) {
+        let entity_id = player.entity_id();
+
+        for player_id in container.all_player_ids() {
+            if entity_id == player_id {
+                container.clear_all_slots().await;
+            }
+        }
+
+        // TODO: should re-add all items to player or drop?
+    }
 }
 
 impl CraftingTableBlock {
     pub async fn open_crafting_screen(
         &self,
+        block: &Block,
         player: &Player,
         location: WorldPosition,
         server: &Server,
     ) {
         //TODO: Adjust /craft command to real crafting table
         let entity_id = player.entity_id();
-        player.open_container.store(Some(1));
-        {
-            let mut open_containers = server.open_containers.write().await;
-            if let Some(ender_chest) = open_containers.get_mut(&1) {
+        let mut open_containers = server.open_containers.write().await;
+        let mut id_to_use = -1;
+
+        for (id, container) in open_containers.iter() {
+            if let Some(a_block) = container.get_block() {
+                if a_block.id == block.id && container.all_player_ids().is_empty() {
+                    id_to_use = *id as i64;
+                }
+            }
+        }
+
+        if id_to_use == -1 {
+            let new_id = server.new_container_id();
+
+            log::info!("New ct ID: {}", new_id);
+
+            let open_container = OpenContainer::new_empty_container::<CraftingTable>(
+                entity_id,
+                Some(location),
+                Some(block.clone()),
+            );
+
+            open_containers.insert(new_id.into(), open_container);
+
+            player.open_container.store(Some(new_id.into()));
+        } else {
+            log::info!("Using previous ct ID: {}", id_to_use);
+            if let Some(ender_chest) = open_containers.get_mut(&(id_to_use as u64)) {
                 ender_chest.add_player(entity_id);
-            } else {
-                let open_container = OpenContainer::new_empty_container::<CraftingTable>(
-                    entity_id,
-                    Some(location),
-                    get_block("minecraft:crafting_table").cloned(),
-                );
-                open_containers.insert(1, open_container);
+                player
+                    .open_container
+                    .store(Some(id_to_use.try_into().unwrap()));
             }
         }
+        drop(open_containers);
```

Comment:
not sure what's the need for manually managing memory as if it was C

Reply:
> not sure what's the need for manually managing memory as if it was C

I think it's not necassary here but it's more like freeing the guard then managing the memory. 
For example if you call an sub function that trys to lock the variable too you create an deadlock.
Which results in freezing

Reply:
I think this is a draft and in progress :D I think I was trying to fix a different issue that I thought was from a deadlock. I will probably refactor this code